format/partition-statistics-spec.md

```diff
@@ -0,0 +1,43 @@
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Partition Statistics file format
+
+This is a specification for partition statistics files. It is designed to store statistics information 
+for every partition value as a row in the **table default format** sorted based on the `partition`.
+
+The schema of the file is as follows:
+
+| Field Id | Field Name                   | Field Type   | isOptional | Doc                                                                                                                         |
+|----------|------------------------------|--------------|------------|-----------------------------------------------------------------------------------------------------------------------------|
+| 1        | partition                    | StructType   | false      | See [PartitionData](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/PartitionData.java) |                                                                                  
+| 2        | spec_id                      | IntegerType  | false      | Partition spec id                                                                                                           |
+| 3        | data_record_count            | LongType     | false      | Count of records in data files                                                                                              |                                               
+| 4        | data_file_count              | IntegerType  | false      | Count of data files                                                                                                         |
+| 5        | position_delete_record_count | LongType     | true       | Count of records in position delete files                                                                                   |                                                          
+| 6        | position_delete_file_count   | IntegerType  | true       | Count of position delete files                                                                                              |                                   
+| 7        | equality_delete_record_count | LongType     | true       | Count of records in equality delete files                                                                                   |                                  
+| 8        | equality_delete_file_count   | IntegerType  | true       | Count of equality delete files                                                                                              |                                
```

Comment:
Why include equality/position delete counts rather than requiring an accurate record count? I don't have an opinion yet either way, but we could require accurate counts here.

Reply:
I think computing the actual count can be a very expensive operation. That too when equality deletes are involved. 
I have followed the same schema that I have implemented for  `Partitions` metadata table.

Reply:
> I think computing the actual count can be a very expensive operation.

I think that's the point. Maybe we should have more accurate data since this can be computed async.

Reply:
If we apply the deletes, maybe the equality_delete and position_delete fields can be empty? If we don't apply the deletes (in some case where we complete stats synchronously during the write in future) it can have the actual delete stats. So, the current schema can still hold good. 

Partitions metadata table is also an async call w.r.t write. But we still don't apply the actual delete files. 
@szehon-ho: Do you have any suggestions?

Reply:
I think it would be reasonable to make this an actual record count for the partition even with the presence of row-level deletes.   Writers that produce positional deletes would be able to produce this count directly during writes.  Equality delete writers like Flink wouldn't be able to produce the actual count, in which case they shouldn't write any partition stats for the snapshot.  Either an async job as suggested, or regularly scheduled compaction jobs, could produce the partition stats in that case.  If partition stats are not present, readers can either choose to use partition stats from earlier snapshots - e.g. from the last compaction - or just fall back to estimates based on other sources (full table rowcount, etc.)

Reply:
@sgcowell: For record count, if we have it as an actual count, then file count should include positional deletes too? Because query planner needs to know the number of expected IO (based on the file counts)?

Reply:
done. Only the accurate count is kept now. Removed delete file stats.

Reply:
> Writers that produce positional deletes would be able to produce this count directly during writes.

This may not be true. For example, for a data file `df1`, we could have two positional delete files(`pdf1`, `pdf2`) applying to it. `pdf1` and `pdf2` may delete the same row in `df1`. The writer may not know if a row has been deleted previously.  Not 100% sure though. Needs inputs from @szehon-ho @aokolnychyi

Reply:
We would need to read all of the valid positional delete files to know the number of unique deletes for a data file.

Reply:
An additional scan on all positional delete files of partition changed are needed. This may still be expensive for sync operations when a write touch a lot of partitions. 

Admittedly, the precise actual record count is useful. I still think the separated metrics(row count, file size) for data files, delete files are valuable. Users could use these metrics to decide if a compaction is needed, or debug why certain scan is slow. For example, if we see the there are several delete files in a partition, we may compact it for better scan perf. With that, I propose to keep them instead of removing.

Reply:
@ajantha-bhat I think file count should remain just the data file count.   Mixing the two reduces the value of the count in my opinion.   Knowing delete file counts isn't that useful in my experience, there's not a lot of choices the planner can make regarding how delete files are handled, even if there are a lot of them.

Reply:
> Partitions metadata table is also an async call w.r.t write. But we still don't apply the actual delete files.
@szehon-ho: Do you have any suggestions?

I still not very convinced that Partitions metadata table should have that column, as it would really reduce the performance of that table.  It's a metadata table, so its sync in the sense that it needs to be calculated when the user queries it.

I slightly prefer having more stats than less. ie total_record_count, data_file_record_count , eq_delete_record_count, pos_delete_record_count.  Internally we can definitely make use of delete record count in planning, for example, like potentially whether to cache deletes, etc.  Users can also use them as @flyrain mentioned.  It is true we already can access all file record counts via manifest entries, so maybe an argument can be made against having them here.  But I think as they are cheap to get, don't see a strong reason why to hide this from the user.

The total_record_count seems the most valuable but expensive to compute.  Also agree with @flyrain  that total count may not be available in writer, for example row-level operation writers , not sure how the writer will know how many total data records are in partition after the write..?